frame/vesting/src/lib.rs

```diff
@@ -312,105 +369,381 @@ pub mod pallet {
+			if let Some(new_schedule) = Self::merge_vesting_info(now, schedule1, schedule2)? {
+				// Merging created a new schedule so we:
+				// 1) need to add it to the accounts vesting schedule collection,
+				schedules.push(new_schedule);
+				// (we use `locked_at` in case this is a schedule that started in the past)
+				let new_schedule_locked = new_schedule.locked_at::<T::BlockNumberToBalance>(now);
+				// 2) update the locked amount to reflect the schedule we just added,
+				locked_now = locked_now.saturating_add(new_schedule_locked);
+				// and 3) deposit an event.
+				Self::deposit_event(Event::<T>::MergedScheduleAdded(
+					new_schedule.locked(),
+					new_schedule.per_block(),
+					new_schedule.starting_block(),
+				));
+			} // In the None case there was no new schedule to account for.
+
+			debug_assert!(
+				locked_now > Zero::zero() && schedules.len() > 0 ||
+					locked_now == Zero::zero() && schedules.len() == 0
+			);
+			Self::write_vesting(&who, schedules)?;
+			Self::write_lock(&who, locked_now);
 
 			Ok(())
 		}
 	}
 }
 
 impl<T: Config> Pallet<T> {
-	/// (Re)set or remove the pallet's currency lock on `who`'s account in accordance with their
-	/// current unvested amount.
-	fn update_lock(who: T::AccountId) -> DispatchResult {
-		let vesting = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+	// Create a new `VestingInfo`, based off of two other `VestingInfo`s.
+	// NOTE: We assume both schedules have had funds unlocked up through the current block.
+	fn merge_vesting_info(
+		now: T::BlockNumber,
+		schedule1: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+		schedule2: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+	) -> Result<Option<VestingInfo<BalanceOf<T>, T::BlockNumber>>, DispatchError> {
+		let schedule1_ending_block = schedule1.ending_block_as_balance::<T::BlockNumberToBalance, T>()?;
+		let schedule2_ending_block = schedule2.ending_block_as_balance::<T::BlockNumberToBalance, T>()?;
+		let now_as_balance = T::BlockNumberToBalance::convert(now);
+
+		// Check if one or both schedules have ended.
+		match (schedule1_ending_block <= now_as_balance, schedule2_ending_block <= now_as_balance) {
+			// If both schedules have ended, we don't merge and exit early.
+			(true, true) => return Ok(None),
+			// If one schedule has ended, we treat the one that has not ended as the new
+			// merged schedule.
+			(true, false) => return Ok(Some(schedule2)),
+			(false, true) => return Ok(Some(schedule1)),
+			// If neither schedule has ended don't exit early.
+			_ => {}
+		}
+
+		let locked = schedule1
+			.locked_at::<T::BlockNumberToBalance>(now)
+			.saturating_add(schedule2.locked_at::<T::BlockNumberToBalance>(now));
+		// This shouldn't happen because we know at least one ending block is greater than now.
+		if locked.is_zero() {
+			log::warn!(
+				target: LOG_TARGET,
+				"merge_vesting_info validation checks failed to catch a locked of 0"
+			);
+			return Ok(None);
+		}
```

Comment:
Should this be a debug assert?

Reply:
Or how could this happen?

Reply:
It would happen if both schedules finish by the current block, which we explicitly check for above. 

It is fine if this happens and we can safely return `Ok(None)`, but it means the code to get here does not behave as we expect.

A debug assert would be helpful, just not sure of an idiomatic way to do it here since we have a log message + a valid return

Reply:
96af8dd13c2beb5473dbf2a4d299af1982d1762e

Reply:
well i mean code doesnt act in magical ways. If the code written above makes this impossible, then it is strange to have this in here

Reply:
>  code doesnt act in magical ways

Code never does, but follow up contributors do :P

In all seriousness tho, what would your rec be? Remove the check and just have a debug assert? 

My thinking here was that if we can be defensive against future mistakes, why not.
But I really don't have enough exprience to have a strong opinion

Reply:
I think a debug assert is better at catching a problem to changed code than a log message which currently is unreachable.

also in genral, we should assume people who update code do so well. there is no amount of 'extra code' you can write which will protect someone who writes bad code, and ultimately can change all of your checks and assumptions.

You should write any assumptions you have as code comments, and use debug assert to verify those assumptions are withheld in all tests